cardano-crypto-tests/src/Test/Crypto/DSIGN.hs

```diff
@@ -11,17 +11,122 @@ module Test.Crypto.DSIGN
   )
 where
 
+import Text.Show.Pretty (ppShow)
+import Control.Monad (replicateM)
+import Data.ByteString (ByteString)
+import qualified GHC.Exts as GHC
+import qualified Test.QuickCheck.Gen as Gen
+import qualified Crypto.Secp256k1 as SECP
+import Data.Kind (Type)
 import Data.Proxy (Proxy (..))
-
-import Cardano.Crypto.DSIGN
-import Cardano.Crypto.Util (SignableRepresentation(..))
-
-import Test.Crypto.Util hiding (label)
+import Cardano.Crypto.DSIGN (
+  MockDSIGN, 
+  Ed25519DSIGN, 
+  Ed448DSIGN,
+  SECP256k1DSIGN,
+  DSIGNAlgorithm (VerKeyDSIGN,
+                  SignKeyDSIGN,
+                  SigDSIGN,
+                  ContextDSIGN,
+                  Signable,
+                  rawSerialiseVerKeyDSIGN,
+                  rawDeserialiseVerKeyDSIGN,
+                  rawSerialiseSignKeyDSIGN,
+                  rawDeserialiseSignKeyDSIGN,
+                  rawSerialiseSigDSIGN,
+                  rawDeserialiseSigDSIGN),
+  sizeVerKeyDSIGN,
+  sizeSignKeyDSIGN,
+  sizeSigDSIGN,
+  encodeVerKeyDSIGN,
+  decodeVerKeyDSIGN,
+  encodeSignKeyDSIGN,
+  decodeSignKeyDSIGN,
+  encodeSigDSIGN,
+  decodeSigDSIGN,
+  signDSIGN,
+  deriveVerKeyDSIGN,
+  verifyDSIGN,
+  genKeyDSIGN,
+  seedSizeDSIGN,
+  )
+import Cardano.Binary (FromCBOR, ToCBOR)
+import Test.Crypto.Util (
+  Message,
+  prop_raw_serialise,
+  prop_size_serialise,
+  prop_cbor_with,
+  prop_cbor,
+  prop_cbor_size,
+  prop_cbor_direct_vs_class,
+  prop_no_thunks,
+  arbitrarySeedOfSize
+  )
 import Test.Crypto.Instances ()
-import Test.QuickCheck ((=/=), (===), (==>), Arbitrary(..), Gen, Property)
+import Test.QuickCheck (
+  (=/=), 
+  (===), 
+  Arbitrary(..), 
+  Gen, 
+  Property,
+  forAllShow
+  )
 import Test.Tasty (TestTree, testGroup)
 import Test.Tasty.QuickCheck (testProperty)
 
+-- Captures ways of generating each of the associated types of DSIGN
+data TypesMethodology (a :: Type) = 
+  TypesMethodology (Gen (VerKeyDSIGN a))
+                   (Gen (SignKeyDSIGN a))
```

Comment:
If we're going to do this, arguably we could get rid of the key generation method from the `DSIGN` class, since it's presumably only used for testing, and this lets us give the generators per-type as we need.

Reply:
I'm happy to make that change, `DSIGN` is already huge.

Reply:
This is potentially quite a breaking change, so I'm reluctant to do this unless absolutely necessary. I've fixed Tobias' original over-generality concern without this, but if you feel a change to `DSIGN` is needed to get this merged, I can do it.

Reply:
Nope, not necessary, it just seems ugly to me to have the generator in there.